Implement cross-origin redirects when loading site-isolated iframes

achristensen07 opened the pull request:
#### a9f7a67a74b27f5af55ea016250f837b008da857
<pre>
Implement cross-origin redirects when loading site-isolated iframes
<a href="https://bugs.webkit.org/show_bug.cgi?id=259817">https://bugs.webkit.org/show_bug.cgi?id=259817</a>
rdar://113377510

Reviewed by Chris Dumez.

In order to get this to work, I needed to do several things:
1. Implement some message handlers in RemotePageProxy to call WebPageProxy methods with the correct process.
2. Make 2 modifications to WebFrameProxy::prepareForProvisionalNavigationInProcess:
2.a. If we are redirected back to the WebFrameProxy&apos;s process, stop using a ProvisionalFrameProxy or RemotePageProxy.
     We are just doing a same-process navigation when this happens.
2.b. If we have a provisional frame but we are doing a cross-origin redirect, make a new provisional frame in the new process.
3. After preparing for a provisional navigation, skip the didCommitLoadForFrame call from the new process
   because we&apos;ve already received it from the first page.  This is what YesAfterProvisionalLoadStarted does.
4. If we are redirecting, frameProcessBeforeNavigation needs to be the provisional frame&apos;s process for continueWithProcessForNavigation.
5. If site isolation is enabled, we can&apos;t reuse a process that hasn&apos;t seen a provisional navigation committed because we
   need a provisional load starting to a new domain to come from a new domain&apos;s process.  This may be optimized away when
   we implement redirect chain following without starting processes for each domain along the chain, but the change
   in WebProcessPool::processForNavigationInternal is needed until then.

Covered by testing a variety of same-domain and cross-domain redirects in addition to existing tests.

* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::load):
* Source/WebKit/UIProcess/API/APINavigation.cpp:
(API::Navigation::currentRequestIsCrossOriginRedirect const):
* Source/WebKit/UIProcess/API/APINavigation.h:
* Source/WebKit/UIProcess/RemotePageProxy.cpp:
(WebKit::RemotePageProxy::didReceiveMessage):
(WebKit::RemotePageProxy::decidePolicyForNavigationActionAsync):
(WebKit::RemotePageProxy::decidePolicyForNavigationActionSync):
(WebKit::RemotePageProxy::didChangeProvisionalURLForFrame):
(WebKit::RemotePageProxy::didReceiveSyncMessage):
* Source/WebKit/UIProcess/RemotePageProxy.h:
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::prepareForProvisionalNavigationInProcess):
* Source/WebKit/UIProcess/WebFrameProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::continueNavigationInNewProcess):
(WebKit::WebPageProxy::decidePolicyForNavigationAction):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::processForNavigationInternal):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/266768@main">https://commits.webkit.org/266768@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2ae6f5c8a133db34b27e7710539afaf29b615040

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/14731 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/15039 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15390 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/16479 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/13891 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/14868 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17559 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/15137 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/16536 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/14911 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15415 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12505 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/17214 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12699 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/13288 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/20285 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13773 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13455 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16699 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/14000 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/11846 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/13298 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17636 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1763 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13844 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/API/APINavigation.cpp:
```diff
@@ -113,6 +114,12 @@ void Navigation::appendRedirectionURL(const WTF::URL& url)
         m_redirectChain.append(url);
 }
 
+bool Navigation::currentRequestIsCrossOriginRedirect() const
```
Based on the implementation, it sounds like you mean cross-site, not cross-origin.

cdumez commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -1947,7 +1947,7 @@ std::tuple<Ref<WebProcessProxy>, SuspendedPageProxy*, ASCIILiteral> WebProcessPo
         return { WTFMove(sourceProcess), nullptr, "An automation session is active"_s };
 
-    if (!sourceProcess->hasCommittedAnyProvisionalLoads()) {
+    if (!sourceProcess->hasCommittedAnyProvisionalLoads() && !page.preferences().siteIsolationEnabled()) {
```
Why are we disabling this optimization? This seems bad. I don't think this optimization is incompatible with site isolation.

achristensen07 replied:
When redirecting multiple times, I was making a new RemotePageProxy before clearing the old one, which caused two RemotePageProxies to be using the same process with the same page ID at the same time.  I'm fixing that by cleaning the old RemotePageProxy before making a new one instead of this.

achristensen07 committed:
Implement cross-origin redirects when loading site-isolated iframes
https://bugs.webkit.org/show_bug.cgi?id=259817
rdar://113377510

Reviewed by Chris Dumez.

In order to get this to work, I needed to do several things:
1. Implement some message handlers in RemotePageProxy to call WebPageProxy methods with the correct process.
2. Make 2 modifications to WebFrameProxy::prepareForProvisionalNavigationInProcess:
2.a. If we are redirected back to the WebFrameProxy's process, stop using a ProvisionalFrameProxy or RemotePageProxy.
     We are just doing a same-process navigation when this happens.
2.b. If we have a provisional frame but we are doing a cross-origin redirect, make a new provisional frame in the new process.
3. After preparing for a provisional navigation, skip the didCommitLoadForFrame call from the new process
   because we've already received it from the first page.  This is what YesAfterProvisionalLoadStarted does.
4. If we are redirecting, frameProcessBeforeNavigation needs to be the provisional frame's process for continueWithProcessForNavigation.
5. If site isolation is enabled, we can't reuse a process that hasn't seen a provisional navigation committed because we
   need a provisional load starting to a new domain to come from a new domain's process.  This may be optimized away when
   we implement redirect chain following without starting processes for each domain along the chain, but the change
   in WebProcessPool::processForNavigationInternal is needed until then.

Covered by testing a variety of same-domain and cross-domain redirects in addition to existing tests.

* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::load):
* Source/WebKit/UIProcess/API/APINavigation.cpp:
(API::Navigation::currentRequestIsCrossOriginRedirect const):
* Source/WebKit/UIProcess/API/APINavigation.h:
* Source/WebKit/UIProcess/RemotePageProxy.cpp:
(WebKit::RemotePageProxy::didReceiveMessage):
(WebKit::RemotePageProxy::decidePolicyForNavigationActionAsync):
(WebKit::RemotePageProxy::decidePolicyForNavigationActionSync):
(WebKit::RemotePageProxy::didChangeProvisionalURLForFrame):
(WebKit::RemotePageProxy::didReceiveSyncMessage):
* Source/WebKit/UIProcess/RemotePageProxy.h:
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::prepareForProvisionalNavigationInProcess):
* Source/WebKit/UIProcess/WebFrameProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::continueNavigationInNewProcess):
(WebKit::WebPageProxy::decidePolicyForNavigationAction):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::processForNavigationInternal):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/266768@main
